### PR TITLE
[ready for review] Hierarchize transforms at cell level

### PR DESCRIPTION
**Pull request by mp3butcher**
following https://github.com/OpenMW/openmw/pull/2421 (for reference)
push hierarchic transforms at cell levels (require testing, perhaps i missed special transform ajustement)

### CODE REVIEW

**Comment by mp3butcher** [comment]
arg the last commit broke stuff

File: apps/openmw/mwrender/objects.cpp [hunk 1]
```diff
@@ -197,9 +213,14 @@ void Objects::updatePtr(const MWWorld::Ptr &old, const MWWorld::Ptr &cur)
...
         }
 
-    if (objectNode->getNumParents())
+    if (objectNode->getNumParents()>0)
+    {
         objectNode->getParent(0)->removeChild(objectNode);
-    cellnode->addChild(objectNode);
+    }
+
+    const float *f = cur.getRefData().getPosition().pos;
```
Comment (Capostrophic) [inline apps/openmw/mwrender/objects.cpp]: Keep in mind getPosition().asVec3() is a thing, and it's not just here you can use it.

File: apps/openmw/mwrender/objects.hpp [hunk 2]
```diff
@@ -6,7 +6,8 @@
 
 #include <osg/ref_ptr>
 #include <osg/Object>
-
+#include <osg/Vec3>
```
Comment (Capostrophic) [inline apps/openmw/mwrender/objects.hpp]: This specific header doesn't use vectors in its declarations. Put this include into files that actually use vectors to keep compilation times shorter.

File: components/sceneutil/util.cpp [hunk 3]
```diff
@@ -1,10 +1,20 @@
 #include "util.hpp"
-
 #include <osg/Node>
+#include <components/esm/loadcell.hpp>
+#include <components/esm/loadland.hpp>
 
 namespace SceneUtil
 {
 
+const float cellSize = static_cast<float>(ESM::Land::REAL_SIZE);
```
Comment (Capostrophic) [inline components/sceneutil/util.cpp]: Global variables are a no-no.

Comment (mp3butcher) [reply components/sceneutil/util.cpp]: any reason for it?

File: apps/openmw/mwrender/objects.cpp (hunk 1 above)
Comment (mp3butcher) [reply apps/openmw/mwrender/objects.cpp]: i just used the original syntax...but if you think it's better, i have nothing against it

File: apps/openmw/mwrender/objects.hpp (hunk 2 above)
Comment (mp3butcher) [reply apps/openmw/mwrender/objects.hpp]: which file do you think about?

Comment (Capostrophic) [reply apps/openmw/mwrender/objects.hpp]: Um. The source files that fail to compile if the line is removed.

**Comment by mp3butcher** [comment]
comply with @Capostophic review

File: apps/openmw/mwrender/renderingmanager.cpp [hunk 4]
```diff
@@ -1152,7 +1153,8 @@ namespace MWRender
     {
         mPlayerAnimation = new NpcAnimation(player, player.getRefData().getBaseNode(), mResourceSystem, 0, NpcAnimation::VM_Normal,
                                                 mFirstPersonFieldOfView);
-
+        if(mPlayerNode->getNumParents()==0)
+            mObjects->getOrCreateCell(player.getCell())->addChild(mPlayerNode);
```
Comment (Capostrophic) [inline apps/openmw/mwrender/renderingmanager.cpp]: This line causes a crash (null pointer dereferencing) for me. Player's cell can and will be null before the game is fully loaded. How did it get past your own testing?

Comment (mp3butcher) [reply apps/openmw/mwrender/renderingmanager.cpp]: newgame was a case i haven't checked..I added a fix...it seems to do the trick

**Comment by akortunov** [comment]
So which gains this PR is supposed to give?

**Comment by mp3butcher** [comment]
> So which gains this PR is supposed to give?
What that must be done must be done....

No hierarchy -> huge transforms on each object leads to numeric problem if you want to flatten or merge them

**Comment by psi29a** [comment]
Close this for now, unless more work is inbound. Obviously as time goes on, we have merge conflicts.

It would be nice if there was an explanation about what this change attempts to do and how it would benefit OpenMW.

Feel free to reopen here: https://gitlab.com/OpenMW/openmw

**Comment by AnyOldName3** [comment]
The main motivation for me suggesting this was that I thought it would be a necessary prerequisite for what was then known as dynamic batching, i.e. what we now call object paging. It shouldn't be necessary any more now object paging is implemented.

**Comment by psi29a** [comment]
It touches a lot of things around light, I'm wondering if was also trying to solve that issue too?

**Comment by Capostrophic** [comment]
If you mean the light limit issue then I'm pretty sure that wasn't the point of changes in lightutil. They make sure a light source's position is adjusted properly taking the cell origin into account since light sources aren't handled like most objects.

**Pull request closed without merging**